### ISSUE TITLE
Build a model docs suggested tweaks

### ISSUE THREAD

**Issue by chrisb13** [ACCESS-NRI/access-hive.org.au#845]
I was following these docs:
https://access-hive.org.au/models/run-a-model/build_a_model/
and found that some of the instructions felt more complicated than necessary, or lacked small useful details.

When I compiled a fresh version of ACCESS OM3, the executables didn't quite match the paths suggested here `## Output directory for compiled packages` (in my case `/g/data/tm70/cyb561/spack/0.22/release/linux-rocky8-x86_64/intel-2021.10.0/access-om3-nuopc-git.0.3.1_0.3.1-va553vmg3iv3i7uknyyymh34nna2hz7s/bin/access-om3-MOM6-CICE6`). 

I also had to guess how to find the hash information (`/g/data/tm70/cyb561/spack/0.22/environments/mom6_dev/spack.lock`). 

Additionally, I think most users will be using the `Spack setup` instructions so I would suggest the text in the green box goes at the top and the edge cases goes later? (Saves the user having to learn about Spack's complicated directory system unnecessarily.)

This statement is important: `The source code is automatically copied inside the environment's folder` but could be more explicit, for me it was: `/g/data/tm70/cyb561/spack/0.22/environments/mom6_dev/access-om3-nuopc`, so adding an example path would help (`/g/data/$PROJECT/$USER/spack/0.22/environments/#ENV_NAME/mom5` I think)

**Comment by chrisb13** [comment]
@anton-seaice also taught me this handy undo process when something goes wrong with one's develop workflow
```bash
spack undevelop access-om3-nuopc
#remove access-om3-nuopc from environments folder
spack -d develop access-om3-nuopc@git.main
```
Would be worth mentioning `undevelop` and `-d` I think

**Comment by anton-seaice** [comment]
We discovered also that its best if `spack develop` uses the same version as the version already in the spec for an environment. 

See notes here about this (and the alternative which uses the `spack change` command) : https://github.com/spack/spack/wiki/Telcon:-2024-12-03

**Comment by atteggiani** [comment]
Hi @chrisb13 and @anton-seaice, 

Thank you for raising the issue and for the information provided.

@chrisb13
> When I compiled a fresh version of ACCESS OM3, the executables didn't quite match the paths suggested here `## Output directory for compiled packages` (in my case `/g/data/tm70/cyb561/spack/0.22/release/linux-rocky8-x86_64/intel-2021.10.0/access-om3-nuopc-git.0.3.1_0.3.1-va553vmg3iv3i7uknyyymh34nna2hz7s/bin/access-om3-MOM6-CICE6`).

The path you specified `/g/data/tm70/cyb561/spack/0.22/release/linux-rocky8-x86_64/intel-2021.10.0/access-om3-nuopc-git.0.3.1_0.3.1-va553vmg3iv3i7uknyyymh34nna2hz7s` (which is the package folder) is in line with the one specified in [Output directory for compiled packages](https://access-hive.org.au/models/run-a-model/build_a_model/#output-directory-for-compiled-packages) (i.e.: `<install_tree.root>/<architecture>/<compiler.name>-<compiler.version>/<name>-<version>-<hash>`. 
 
> I also had to guess how to find the hash information (`/g/data/tm70/cyb561/spack/0.22/environments/mom6_dev/spack.lock`).

There is no information about the hash structure and I agree some information might be useful.
@harshula I am not familiar with `spack` hashes, do we have any information that might be useful to add?

> Additionally, I think most users will be using the `Spack setup` instructions so I would suggest the text in the green box goes at the top and the edge cases goes later? (Saves the user having to learn about Spack's complicated directory system unnecessarily.)

We could do it.
However, the instructions might still be not identical to all user-case scenarios, because of the Spack's complicated directory/scope system (and because we might be changing environment/packages specifications over time). Since we cannot provide instructions that reflect 100% what a user will see, I think the best we can do is trying to provide more information, in case what users see doesn't exactly match what the documentation says.
What type of information do you think the page lacks at the moment and could be added to make the life of a user easier?

> This statement is important: `The source code is automatically copied inside the environment's folder` but could be more explicit, for me it was: `/g/data/tm70/cyb561/spack/0.22/environments/mom6_dev/access-om3-nuopc`, so adding an example path would help (`/g/data/$PROJECT/$USER/spack/0.22/environments/#ENV_NAME/mom5` I think)

Something about that might have been done within [this spack develop PR](https://github.com/ACCESS-NRI/access-hive.org.au/pull/888).
Please feel free to review it.

> [@anton-seaice](https://github.com/anton-seaice) also taught me this handy undo process when something goes wrong with one's develop workflow
> 
> spack undevelop access-om3-nuopc
> #remove access-om3-nuopc from environments folder
> spack -d develop access-om3-nuopc@git.main
> Would be worth mentioning `undevelop` and `-d` I think

I agree `spack undevelop` is a useful command. 
We could add a paragraph about it in the documentation.

What does `spack -d ...` exactly do? I couldn't find any documentation on the spack website.

@anton-seaice 
We discovered also that its best if spack develop uses the same version as the version already in the spec for an environment.

Something along that line is included within the spack develop PR mentioned above.
Please feel free to have a look and review it.

**Comment by chrisb13** [comment]
Thanks @atteggiani for the detailed follow up. This was all a while ago but I'll do my best...

> The path you specified `/g/data/tm70/cyb561/spack/0.22/release/linux-rocky8-x86_64/intel-2021.10.0/access-om3-nuopc-git.0.3.1_0.3.1-va553vmg3iv3i7uknyyymh34nna2hz7s` (which is the package folder) is in line with the one specified in [Output directory for compiled packages](https://access-hive.org.au/models/run-a-model/build_a_model/#output-directory-for-compiled-packages) (i.e.: `<install_tree.root>/<architecture>/<compiler.name>-<compiler.version>/<name>-<version>-<hash>`.

From memory it was the `restricted` part of this path that I found to be different:
`/g/data/$PROJECT/$USER/spack/0.22/spack/../restricted/ukmo/release/<architecture>/<compiler.name>-<compiler.version>/<name>-<version>-<hash>`
Maybe that comes under `architecture` above?

> Additionally, I think most users will be using the `Spack setup` instructions so I would suggest the text in the green box goes at the top and the edge cases goes later? (Saves the user having to learn about Spack's complicated directory system unnecessarily.)
...
What type of information do you think the page lacks at the moment and could be added to make the life of a user easier?

I'm not sure if my suggestion last time was very easy to follow, let me try again: I'm thinking this text:

>For the Spack instance obtained through the [Spack setup instructions](https://access-hive.org.au/getting_started/spack), $spack (referred to as $(prefix) in Spack configuration scopes documentation above) corresponds to the /g/data/$PROJECT/$USER/spack/0.22/spack directory.

Could go just under `Output directory for compiled packages`, then after the green box you could have the more nuanced text with an explanation that Spack complications and changing the install path might change the actual location. 

> > This statement is important: `The source code is automatically copied inside the environment's folder` but could be more explicit, for me it was: `/g/data/tm70/cyb561/spack/0.22/environments/mom6_dev/access-om3-nuopc`, so adding an example path would help (`/g/data/$PROJECT/$USER/spack/0.22/environments/#ENV_NAME/mom5` I think)
> 
> Something about that might have been done within [this spack develop PR](https://github.com/ACCESS-NRI/access-hive.org.au/pull/888). Please feel free to review it.

H'm from what I can [tell it's been deleted]()?

**Comment by chrisb13** [comment]
On a broader point, I think we discussed online @atteggiani ages ago? I find the naming of this page a little narrow and I imagine reduces the broad appeal. Would it not be better to re-name to something like:
"How to build and/or modify an ACCESS model's source code"

There are users who would like to customise a build (e.g. variants/compiler flags etc) but not change source code. With the above in mind, the page could be split into two parts. First building and then (only) if needed modifying/changing build components/flags etc.

Perhaps that kind of change could be considered in [the PR](https://github.com/ACCESS-NRI/access-hive.org.au/pull/888/files) @heidinett?

ping'ing: @helenmacdonald @ezhilsabareesh8 as they've been thinking about similar things ([example](https://forum.access-hive.org.au/t/building-access-om3-with-ww3-using-spack-for-local-modifications-of-ww3/4291)).

**Comment by atteggiani** [comment]
> From memory it was the `restricted` part of this path that I found to be different: `/g/data/$PROJECT/$USER/spack/0.22/spack/../restricted/ukmo/release/<architecture>/<compiler.name>-<compiler.version>/<name>-<version>-<hash>` Maybe that comes under `architecture` above?

The `../restricted/ukmo/release` is set within the `config.install_tree.root` of the example's `spack.yaml` file, as specified within the [documentation](https://access-hive.org.au/pr-preview/pr-888/models/run-a-model/build_a_model/#output-directory-for-compiled-packages).

> The green box (n.d.r.) could go just under `Output directory for compiled packages`, then after the green box you could have the more nuanced text with an explanation that Spack complications and changing the install path might change the actual location.

Yes, I agree the green box specified right after the paragraph title can be of help.
I will update the [modify spack develop PR](https://github.com/ACCESS-NRI/access-hive.org.au/pull/888) with this change.

> H'm from what I can [tell it's been deleted](https://github.com/ACCESS-NRI/access-hive.org.au/pull/888/files#r1988359579)?

I replied within [the related comment](https://github.com/ACCESS-NRI/access-hive.org.au/pull/888#discussion_r1988371312) of the modify spack develop PR.

**Comment by atteggiani** [comment]
All good. We will leave this issue as `Parked` as we don't have time to work on this at the moment.

**Comment by harshula** [comment]
Hi @chrisb13 , When we migrate to Spack v1.x, this document will need some important changes. Perhaps we can make these changes at the same time?

FYI, The document has moved to: https://docs.access-hive.org.au/models/build_a_model/build_source_code/

**Comment by chrisb13** [comment]
> When we migrate to Spack v1.x, this document will need some important changes. Perhaps we can make these changes at the same time?

Sounds good.

**Comment by chrisb13** [comment]
Just noting from @harshula and Chris chat today. 

The COSIMA community currently come from the past experience of building OM2 with this:

>cd $ACCESS_OM2_DIR
./install.sh

[source](https://github.com/COSIMA/access-om2/wiki/Getting-started#building-the-models).

So they have high expectations for how easy this is to do...
